lib/diaspora_federation/entities/relayable.rb

```diff
@@ -6,8 +6,50 @@ def self.included(model)
           property :parent_guid
           property :parent_author_signature
           property :author_signature
+
+          # Generates XML and updates signatures
+          def to_xml
+            xml = entity_xml
+            hash = Salmon::XmlPayload.hash(xml)
+            Relayable.update_signatures!(hash)
+
+            xml.at_xpath("author_signature").content = hash[:author_signature]
+            xml.at_xpath("parent_author_signature").content = hash[:parent_author_signature]
+            xml
+          end
         end
       end
+
+      class SignatureVerificationFailed < ArgumentError
+      end
+
+      def self.verify_signatures(data)
+        pkey = DiasporaFederation.callbacks.trigger(:get_public_key_by_id, data[:diaspora_id])
+        raise SignatureVerificationFailed, "failed to fetch public key for #{data[:diaspora_id]}" if pkey.nil?
+        raise SignatureVerificationFailed, "wrong author_signature" unless Signing.verify_signature(
+          data, data[:author_signature], pkey
+        )
+
+        unless DiasporaFederation.callbacks.trigger(:post_author_is_local, data[:parent_guid])
+          # this happens only on downstream federation
+          pkey = DiasporaFederation.callbacks.trigger(:get_public_key_by_post_guid, data[:parent_guid])
+          raise SignatureVerificationFailed,
+                "failed to fetch public key for parent of #{data[:parent_guid]}" if pkey.nil?
+          raise SignatureVerificationFailed, "wrong parent_author_signature" unless Signing.verify_signature(
+            data, data[:parent_author_signature], pkey
+          )
+        end
+      end
+
+      def self.update_signatures!(data)
+        pkey = DiasporaFederation.callbacks.trigger(:get_private_key_by_id, data[:diaspora_id])
+        data[:author_signature] = Signing.sign_with_key(data, pkey) unless pkey.nil?
```

Comment:
And we can't allow ```author_signature``` to be empty, because if we receive an entity from other pod, we don't want it to pass if it isn't filled. But we could though rely on the idea, that entity would fail verification of signature if it's empty and don't check it in the validator.